apps/namadillo/src/lib/query.ts

```diff
@@ -0,0 +1,195 @@
+import invariant from "invariant";
+import { GasConfig } from "types/fees";
+
+const {
+  NAMADA_INTERFACE_NAMADA_TOKEN:
+    nativeToken = "tnam1qxgfw7myv4dh0qna4hq0xdg6lx77fzl7dcem8h7e",
+} = process.env;
+
+export type TransactionPair<T> = {
+  encodedTxData: EncodedTxData<T>;
+  signedTx: Uint8Array;
+};
+
+export type EncodedTxData<T> = {
+  type: string;
+  encodedTx: EncodedTx;
+  meta?: {
+    props: T;
+  };
+};
+
+export type TransactionNotification = {
+  success?: { title: string; text: string };
+  error?: { title: string; text: string };
+};
+
+export type PreparedTransaction = {
+  encodedTx: EncodedTx;
+  signedTx: Uint8Array;
+  notification?: TransactionNotification;
+};
+
+export const revealPublicKeyType = "revealPublicKey";
+
+const getTxProps = (
+  account: Account,
+  gasConfig: GasConfig,
+  chain: Chain
+): WrapperTxMsgValue => {
+  const address = nativeToken;
+  invariant(!!address, "Invalid currency address");
+  invariant(
+    !!account.publicKey,
+    "Account doesn't contain a publicKey attached to it"
+  );
+
+  return {
+    token: address!,
+    feeAmount: gasConfig.gasPrice,
+    gasLimit: gasConfig.gasLimit,
+    chainId: chain.chainId,
+    publicKey: account.publicKey!,
+    memo: "",
+  };
+};
+
+/**
+ * Builds an array of encoded transactions based on the provided query properties.
+ * Each transaction is processed through the provided transaction function `txFn`.
+ * @param {T[]} queryProps - An array of properties used to build transactions.
+ * @param {(WrapperTxProps, T) => Promise<EncodedTx>} txFn - Function to build each transaction.
+ */
+export const buildTxArray = async <T>(
+  account: Account,
+  gasConfig: GasConfig,
+  chain: Chain,
+  queryProps: T[],
+  txFn: (wrapperTxProps: WrapperTxProps, props: T) => Promise<EncodedTx>
+): Promise<EncodedTxData<T>[]> => {
+  const { tx, rpc } = await getSdkInstance();
+  const wrapperTxProps = getTxProps(account, gasConfig, chain);
+  const txArray: EncodedTxData<T>[] = [];
+
+  // Determine if RevealPK is needed:
+  const pk = await rpc.queryPublicKey(account.address);
+  if (!pk) {
+    const revealPkTx = await tx.buildRevealPk(wrapperTxProps, account.address);
+    txArray.push({ type: revealPublicKeyType, encodedTx: revealPkTx });
+  }
+
+  const encodedTxs = await Promise.all(
+    queryProps.map((props) => txFn.apply(tx, [wrapperTxProps, props]))
+  );
+
+  const typedEncodedTxs = encodedTxs.map((encodedTx, index) => ({
+    encodedTx,
+    type: txFn.name,
+    meta: {
+      props: queryProps[index],
+    },
+  }));
+
+  txArray.push(...typedEncodedTxs);
+  return txArray;
+};
+
+/**
+ * Asynchronously signs an array of encoded transactions using Namada extension.
+ */
+export const signTxArray = async <T>(
+  chain: Chain,
+  typedEncodedTxs: EncodedTxData<T>[],
+  owner: string
+): Promise<Uint8Array[]> => {
+  const integration = getIntegration(chain.id);
+  const signingClient = integration.signer() as Signer;
```

Comment:
I wonder why we need this `as Signer`. Not a big deal to leave it for now, just curious!

Reply:
It's partly because of a stupid way I refactored the types so `getIntegration` is returning `unknown`, but if we explicitly pass `"namada"` it won't return `unknown`. But even if we do that it will still return `Signer | undefined` I think.